sdks/java/io/iceberg/src/main/java/org/apache/beam/sdk/io/iceberg/SerializableDataFile.java

```diff
@@ -0,0 +1,206 @@
+ * #createDataFile(PartitionSpec)} to reconstruct the original {@link DataFile}.
+ */
+@DefaultSchema(AutoValueSchema.class)
+@AutoValue
+abstract class SerializableDataFile {
+  public static Builder builder() {
+    return new AutoValue_SerializableDataFile.Builder();
+  }
+
+  abstract String getPath();
+
+  abstract String getFileFormat();
+
+  abstract long getRecordCount();
+
+  abstract long getFileSizeInBytes();
+
+  abstract String getPartitionPath();
+
+  abstract int getPartitionSpecId();
+
+  abstract @Nullable ByteBuffer getKeyMetadata();
+
+  abstract @Nullable List<Long> getSplitOffsets();
+
+  abstract @Nullable Map<Integer, Long> getColumnSizes();
+
+  abstract @Nullable Map<Integer, Long> getValueCounts();
+
+  abstract @Nullable Map<Integer, Long> getNullValueCounts();
+
+  abstract @Nullable Map<Integer, Long> getNanValueCounts();
+
+  abstract @Nullable Map<Integer, byte[]> getLowerBounds();
+
+  abstract @Nullable Map<Integer, byte[]> getUpperBounds();
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setPath(String path);
+
+    abstract Builder setFileFormat(String fileFormat);
+
+    abstract Builder setRecordCount(long recordCount);
+
+    abstract Builder setFileSizeInBytes(long fileSizeInBytes);
+
+    abstract Builder setPartitionPath(String partitionPath);
+
+    abstract Builder setPartitionSpecId(int partitionSpec);
+
+    abstract Builder setKeyMetadata(ByteBuffer keyMetadata);
+
+    abstract Builder setSplitOffsets(List<Long> splitOffsets);
+
+    abstract Builder setColumnSizes(Map<Integer, Long> columnSizes);
+
+    abstract Builder setValueCounts(Map<Integer, Long> valueCounts);
+
+    abstract Builder setNullValueCounts(Map<Integer, Long> nullValueCounts);
+
+    abstract Builder setNanValueCounts(Map<Integer, Long> nanValueCounts);
+
+    abstract Builder setLowerBounds(Map<Integer, byte[]> lowerBounds);
+
+    abstract Builder setUpperBounds(Map<Integer, byte[]> upperBounds);
+
+    abstract SerializableDataFile build();
+  }
+
+  /**
+   * Create a {@link SerializableDataFile} from a {@link DataFile} and its associated {@link
+   * PartitionKey}.
+   */
+  static SerializableDataFile from(DataFile f, PartitionKey key) {
+    SerializableDataFile.Builder builder =
+        SerializableDataFile.builder()
+            .setPath(f.path().toString())
+            .setFileFormat(f.format().toString())
+            .setRecordCount(f.recordCount())
+            .setFileSizeInBytes(f.fileSizeInBytes())
+            .setPartitionPath(key.toPath())
+            .setPartitionSpecId(f.specId())
+            .setKeyMetadata(f.keyMetadata())
+            .setSplitOffsets(f.splitOffsets())
+            .setColumnSizes(f.columnSizes())
+            .setValueCounts(f.valueCounts())
+            .setNullValueCounts(f.nullValueCounts())
+            .setNanValueCounts(f.nanValueCounts());
+
+    // ByteBuddyUtils has trouble converting Map value type ByteBuffer
```

Comment:
Should we be fixing "ByteBuddyUtils" ? At least we should move this logic to a function and add appropriate unit testing.

Reply:
Fair point. The ByteBuddy and type conversion code is a little opaque to me but I'll give it a shot today. If I don't get far, I'll put it in a function

Reply:
Investigated for some time and I think I narrowed it down, but I still am not close to finding a solution. I opened a github issue for this: #32701 

To move forward and get this in for the release, I'll just create a function and reference this issue